File: 0498-diagonal-traverse/0498-diagonal-traverse.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        map<int, vector<int>>mp;
        int n = mat.size(), m = mat[0].size();
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                mp[i + j].push_back(mat[i][j]);
            }
        }

        vector<int>ans;

        for(int i=0; i<=m+n-2; i++){
            if(i % 2){
                vector<int>v = mp[i];
                for(auto el: v) ans.push_back(el);
            }
            else{
                vector<int>v = mp[i];
                reverse(v.begin(), v.end());
                for(auto el: v) ans.push_back(el);
            }
        }
        return ans;
    }
};

File: 0498-diagonal-traverse/0498-diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:

        ans = []
        n = len(mat)
        m = len(mat[0])
        mp = defaultdict(lambda: [])

        for i in range(n):
            for j in range(m):
                mp[i + j].append(mat[i][j])

        for i in mp:
            temp = []
            for el in mp[i]:
                temp.append(el)

            if i % 2 == 0:
                temp.reverse()

            for el in temp:
                ans.append(el)

        return ans
